### DISCUSSION TITLE
Pagination on Variant types.

### DISCUSSION THREAD

**Discussion by supriya-reddy-91** [hayes/pothos#1291] [Q&A] [answered]
Hi team,

We are trying to extend our custom pagination (detailed https://github.com/hayes/pothos/discussions/1206) to all the [`variants`](https://www.npmjs.com/package/@pothos/plugin-prisma#type-variants) on a model.

Our paginated structure has been modified to -

```
export type PaginatedDataShape<T> = {
  take: number;
  skip: number;
  count: number;
  data: T[];
};

// User type is from the prisma-generated types
export const PaginatedUser = builder.objectRef<PaginatedDataShape<User>>('PaginatedUser');

builder.objectType(PaginatedUser, {
  description: 'Paginated version of the User type',
  fields: t => ({
    take: t.exposeInt('take'),
    skip: t.exposeInt('skip'),
    count: t.exposeInt('count'),
    data: t.prismaField({
      type: ['User'],
      resolve: async (_query, parent) => {
        return parent.data;
      },
    }),
  }),
});
``` 

On extending this concept to a variant on the user model, it works when we pass the Primary variant/base model as the object ref.

```
// We expect to pass UserVariant as the objectRef here 
export const PaginatedUserVariant = builder.objectRef<PaginatedDataShape<User>>('PaginatedUserVariant');

builder.objectType(PaginatedUserVariant, {
   description: 'Paginated version of the UserVariant type',
   fields: t => ({
     take: t.exposeInt('take'),
     skip: t.exposeInt('skip'),
     count: t.exposeInt('count'),
     data: t.prismaField({
       type: [UserVariant],
       resolve: async (_query, parent) => {
         return parent.data;
       },
     }),
   }),
 });
 
 ```
 
**Concern/Question** - Could you let us know if this is the right approach? If yes, can you please explain why the object ref takes in the primary variant and returns an additional variant (without the type checker complaining).

Thank you in advance for you time and help.

**Comment by hayes** [comment, marked as answer]
It is likely that the data requirements for both variants are the same.  Typescript uses structural typing, so if 2 types have all the same properties, they can be used interchangeably without any type errors.

I'd have to see the definitions of `UserVariant` and `User` to know for sure
